crates/bevy_ecs/src/system/system_param.rs

```
@@ -752,22 +754,28 @@ all_tuples_enumerated!(impl_param_set, 1, 8, P, p);
 // SAFETY: Res only reads a single World resource
 unsafe impl<'a, T: Resource> ReadOnlySystemParam for Res<'a, T> {}
 
-// SAFETY: Res ComponentId access is applied to SystemMeta. If this Res
-// conflicts with any prior access, a panic will occur.
+// SAFETY: ResMut component access is tied to the singleton components under
+// QueryState<Ref<T>, With<Internal>>, which guarantees safety.
 unsafe impl<'a, T: Resource> SystemParam for Res<'a, T> {
```

Can we add a test that this works "correctly" when a resource requires another resource? E.g. `ResA` requires `ResB`, then you insert `ResA` and try to access `Res<ResB>`. This should return an error because `ResB` should not exist in the world as a resource.

---

For that I need to check whether or not `ResB`s component id is in `resource_entities` or not, which I don't know how to do without using unsound code (previous atttempt you commented on).

Unless you know a relatively easy way to do it, I think it's probably best if we forbid using `require(Resource)` on a resource untill we figure out all of the semantics.

---

> Unless you know a relatively easy way to do it, I think it's probably best if we forbid using require(Resource) on a resource untill we figure out all of the semantics.

That doesn't seem easy to do. You would have to introduce even more hacks in the derive macro (already the fact it calls the derive for `Component` is very hacky!). You would probably have to forbid `require(Component)` in general, not just `require(Resource)`. Moreover users can register required components at runtime, and that seems even harder to prevent.

The way I see it this is a consequence of this PR trying to do two things at the same time:
- store resources as components on entities
- make resources actual components

IMO we should avoid doing the second point until we figure out all its implications. As long as it's not required for the first part it seems just an added complication.

---

I agree, but supporting `#[require(Foo)]` for resources is specifically necessary for this PR to merge.

This is due to the fact `AmbientLight` in `bevy_light` is used as both a `Component` and a `Resource`, requiring `Camera` in the case it is a component.

This should both be a short-lived problem, since the rendering folks have ensured me that `AmbientLight` will be deprecated in 0.18, but for now `Resource`s need to support `require`.

---

> Yes, but in order to store resources on the world, they have to be components

You can just wrap them in a newtype that implemente `Component` and store that in the world.

> Btw, is this the last problem you see with this PR? 

There were a bunch of comments that you marked as solved without any discussion or evident fix. Maybe I missed the change that fixes them, but it's not simple and quite time-expensive to look for them so it would be helpful if you commented on them with why/how they were resolved.